goas/cl/ledger/src/death_constraint.rs

```diff
@@ -0,0 +1,83 @@
+use proof_statements::death_constraint::DeathConstraintPublic;
+use sha2::{Digest, Sha256};
+
+use crate::error::Result;
+
+pub type Risc0DeathConstraintId = [u32; 8];
+
+#[derive(Debug, Clone)]
+pub struct DeathProof {
+    constraint: Risc0DeathConstraintId,
+    risc0_receipt: risc0_zkvm::Receipt,
+}
+
+fn risc0_id_to_cl_death_constraint(risc0_id: Risc0DeathConstraintId) -> [u8; 32] {
+    // RISC0 proof ids have the format: [u32; 8], and cl death constraint ids have the format [u8; 32].
+    // CL death constraints are meaningless beyond being binding, therefore we merely need a collision
+    // resisitant mapping of RISC0 ids to cl death constraints.
+
+    let mut hasher = Sha256::new();
+    hasher.update(b"NOMOS_RISC0_ID_TO_CL_DEATH_CONSTRAINT");
+    for word in risc0_id {
+        hasher.update(u32::to_ne_bytes(word));
+    }
+    let death_constraint: [u8; 32] = hasher.finalize().into();
+    death_constraint
+}
```

Comment:
can't we just re-interpret the bytes? It's 32 bytes anyway

Reply:
yes, that's how I had it initially, but I think it's just a pun that the risc0 stark proof ids happen to be 32 bytes. When we switch to groth16, the constraint vk will not be 32 bytes, and instead we'll need to hash the vk in the same way to bind the note to this vk.

Reply:
Agree, ideally, all proofs should be groth16 even in the PoC?

Reply:
yeah, ideally. Hopefully its a straight forward change to swap out the stark proofs for groth16 proofs.


Maybe we still want to maintain a full stark version of the CL? That way thin clients can build stark proofs and then executors will do the expensive wrapping to groth16?